darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py

```diff
@@ -0,0 +1,442 @@
+                'hostname': 'sn176.localdomain',
+                'write_count': 1,
+                'read_count': 0,
+                'write_segments':
+                    offset  length  start_time  end_time
+                    0       0      40    0.103379  0.103388,
+                'read_segments':
+                    Empty DataFrame
+                    Columns: []
+                    Index: []
+            },
+            {
+                'id': 9457796068806373448,
+                'rank': 0,
+                'hostname': 'sn176.localdomain',
+                'write_count': 1,
+                'read_count': 0,
+                'write_segments':
+                    offset  length  start_time  end_time
+                    0       0    4000    0.104217  0.104231,
+                'read_segments':
+                    Empty DataFrame
+                    Columns: []
+                    Index: []
+            },
+        ]
+
+        (read_df, write_df) = (
+            Empty DataFrame
+            Columns: []
+            Index: [],
+            length  start_time  end_time  rank
+            0      40    0.103379  0.103388     0
+            1    4000    0.104217  0.104231     0
+        )
+
+    """
+    # columns to drop when accumulating the dataframes.
+    # Currently "offset" data is not utilized
+    drop_columns = ["offset"]
+    # create empty arrays to store
+    # read/write segment dataframes
+    read_df_list = []
+    write_df_list = []
+    # iterate over all records/dictionaries
+    # to pull out the dataframes
+    for _dict in dict_list:
+        # collect the read and write segment dataframes
+        rd_seg_df = _dict["read_segments"]
+        wr_seg_df = _dict["write_segments"]
+
+        if rd_seg_df.size:
+            # drop unused columns from the dataframe
+            rd_seg_df = rd_seg_df.drop(columns=drop_columns)
+            # create new column for the ranks
+            rd_seg_df["rank"] = _dict["rank"]
+            # add the dataframe to the list
+            read_df_list.append(rd_seg_df)
+
+        if wr_seg_df.size:
+            # drop unused columns from the dataframe
+            wr_seg_df = wr_seg_df.drop(columns=drop_columns)
+            # create new column for the ranks
+            wr_seg_df["rank"] = _dict["rank"]
+            # add the dataframe to the list
+            write_df_list.append(wr_seg_df)
+
+    if read_df_list:
+        # concatenate the list of pandas dataframes into
+        # a single one with new row indices
+        read_df = pd.concat(read_df_list, ignore_index=True)
+    else:
+        # if the list is empty assign an empty dataframe
+        read_df = pd.DataFrame()
+
+    if write_df_list:
+        # concatenate the list of pandas dataframes into
+        # a single one with new row indices
+        write_df = pd.concat(write_df_list, ignore_index=True)
+    else:
+        # if the list is empty assign an empty dataframe
+        write_df = pd.DataFrame()
```

Comment:
We could probably avoid the duplicated logic for reading and writing data structures I suspect, maybe with a loop if the data structures do indeed actually need to remain separate.

Reply:
The new version refactors this function removing this repeated code. I'll resolve this.